File: hw/fye/code/julia/g-prior.jl
using Distributions

#=
  For the model:
  y | X, β, ϕ  ~ N( Xβ, 1/ϕ Iₙ)
        p(β,ϕ) ~ N(  0, g/ϕ (X'X)⁻¹ )

  By default, g is set to n (the number of observations).
  Note that g-priors work fine for n ≤ p

=#

function gprior(y, X, B; g=0, add_intercept=true, setseed=0)
  if setseed>0 srand(setseed) end
  X = add_intercept ? [fill(1,size(X)[1]) X] : X

  n,k = size(X)
  if g == 0  g = n end

  XX = X'X
  XXi = inv(XX)
  beta_hat = XXi * X'y
  c = g / (1+g)
  sse = sum( (y-X*beta_hat).^2 )
  phi_shape = (n-1)/2
  phi_rate = (sse./2 + beta_hat'*XX*beta_hat / (2*(1+g)))[1]

  samp_beta(phi) = rand(MultivariateNormal(c * beta_hat, c/phi * XXi))
  samp_phi(N) = rand(Gamma( phi_shape, 1/phi_rate ),N)

  new_phi = samp_phi(B)
  tmp_beta = [samp_beta(i) for i in new_phi]
  new_beta = hcat([x for x in tmp_beta]...)'

  Dict(:phi=>new_phi, :beta=>new_beta)
end

File: hw/fye_scala/src/main/scala-2.11/Gprior.scala
package fyeScala

object Gprior {
  import breeze.linalg.{DenseMatrix=>Dmat, DenseVector=>Dvec, inv}
  import breeze.stats.distributions.{MultivariateGaussian=>MVN,Gamma}
  import breeze.numerics.{pow}

  def sample(y: Dvec[Double], X: Dmat[Double], B: Int, gSetter:Double=0, addIntcpt:Boolean=true): 
    (Dvec[Double], Dmat[Double]) = {
    val g = if (gSetter == 0.0) X.rows.toDouble else gSetter

    val (n,k) = (X.rows,X.cols)
    val XX = X.t * X
    val XXi = inv(XX)
    val Xy = X.t * y
    val bHat = XX \ Xy
    val c = g / (1+g)

    val sse = pow(y-X*bHat,2).sum

    val phiShape = (n-1)/2.0
    val phiRate = (sse/2.0 + bHat.t *XX* bHat / (2.0*(1+g)))

    def sampBeta(phi: Double): Dvec[Double] =  MVN(c*bHat, c/phi*XXi).sample
    def sampPhi(N: Int): Dvec[Double] = Gamma(phiShape, 1/phiRate).samplesVector(N)

    val newPhi = sampPhi(B)
    val tmpBeta = newPhi.map(sampBeta(_))
    val newBeta = Dmat.tabulate(tmpBeta.length,tmpBeta(0).length){(i,j) => tmpBeta(i)(j)} 
    //Map("phi" -> newPhi, "beta" -> newBeta)
    (newPhi, newBeta)
  }
}
